python/mscclpp/language/program.py

```diff
@@ -108,20 +112,100 @@ def __init__(
 
         self.loop_context = None
 
+    def __init__(self, spec: AlgoSpec):
+        """Initialize a new CollectiveProgram from an algorithm specification.
+
+        This constructor provides an alternative way to create a CollectiveProgram
+        using an AlgoSpec object, which contains the complete algorithm specification
+        including collective type, protocol parameters, and optimization settings.
+        The collective operation is automatically instantiated based on the collective_name
+        specified in the spec.
+
+        Args:
+            spec (AlgoSpec): Algorithm specification containing all program parameters
+                and configuration settings.
+
+        Raises:
+            AssertionError: If protocol is not "Simple" or "LL".
+            ValueError: If collective_name is not supported.
+
+        Example:
+            >>> from mscclpp.language.utils import AlgoSpec
+            >>> spec = AlgoSpec(
+            ...     name="my_allreduce",
+            ...     collective_name="allreduce",
+            ...     world_size=4,
+            ...     instances=1,
+            ...     protocol="Simple",
+            ...     in_place=False
+            ... )
+            >>> with CollectiveProgram(spec) as prog:
+            ...     # Define communication operations
+            ...     pass
+        """
+        collective = self._create_collective_from_name(spec.collective_name, spec.world_size, spec.in_place)
```

Comment:
Maybe we can let user pass the collective to spec? Not create via name. For example I have two allreduce for size from 1K-1M, one for MI300x one for H100, we need to distinguish them by name